lib/dockerregistry/inventory.go

```diff
@@ -1531,6 +1572,152 @@ func (sc *SyncContext) GarbageCollect(
+		return cr.DockerManifestSchema1, nil
+	case cr.DockerManifestSchema1Signed:
+		return cr.DockerManifestSchema1Signed, nil
+	case cr.DockerManifestSchema2:
+		return cr.DockerManifestSchema2, nil
+	default:
+		return cr.MediaType(""), fmt.Errorf("unsupported MediaType %s", v)
+	}
+}
+
+// ClearRepository wipes out all Docker images from a registry! Use with caution.
+// nolint[gocyclo]
+//
+// TODO: Maybe split this into 2 parts, so that each part can be unit-tested
+// separately (deletion of manifest lists vs deletion of other media types).
+func (sc *SyncContext) ClearRepository(
+	regName RegistryName,
+	mfest Manifest,
+	mkProducer func(RegistryContext, ImageName, Digest) stream.Producer,
+	customProcessRequest *ProcessRequest) {
+
+	// deleteRequstsPopulator returns a PopulateRequests that
+	// varies by a predicate. Closure city!
+	var deleteRequstsPopulator func(func(cr.MediaType) bool) PopulateRequests = func(predicate func(cr.MediaType) bool) PopulateRequests {
+
+		var populateRequests PopulateRequests = func(
+			sc *SyncContext,
+			reqs chan<- stream.ExternalRequest,
+			wg *sync.WaitGroup) {
+
+			for _, registry := range mfest.Registries {
+				// Skip over any registry that does not match the regName we want to
+				// wipe.
+				if registry.Name != regName {
+					continue
+				}
+				for imageName, digestTags := range sc.Inv[registry.Name] {
+					for digest, _ := range digestTags {
+						mediaType, ok := sc.DigestMediaType[digest]
+						if !ok {
+							fmt.Println("could not detect MediaType of digest", digest)
+							continue
+						}
+						if !predicate(mediaType) {
+							fmt.Printf("skipping digest %s mediaType %s\n", digest, mediaType)
+							continue
+						}
+						var req stream.ExternalRequest
+						req.StreamProducer = mkProducer(
+							registry,
+							imageName,
+							digest)
+						req.RequestParams = PromotionRequest{
+							Delete,
+							"",
+							registry.Name,
+							registry.ServiceAccount,
+
+							// No source image name, because tag deletions
+							// should only delete the what's in the
+							// destination registry
+							ImageName(""),
+
+							imageName,
+							digest,
+							"",
+							"",
+						}
+						wg.Add(1)
+						reqs <- req
+					}
+				}
+			}
+		}
+		return populateRequests
+	}
+
+	var processRequest ProcessRequest
+	var processRequestReal ProcessRequest = func(
+		sc *SyncContext,
+		reqs chan stream.ExternalRequest,
+		requestResults chan<- RequestResult,
+		wg *sync.WaitGroup,
+		mutex *sync.Mutex) {
+
+		for req := range reqs {
+			reqRes := RequestResult{Context: req}
+			jsons, errors := getJSONSFromProcess(req)
+			if len(errors) > 0 {
+				reqRes.Errors = errors
+				requestResults <- reqRes
+				// Don't skip over this request, because stderr here is ignorable.
+				// continue
+			}
+			for _, json := range jsons {
+				sc.Info("DELETED image:", json)
+			}
+			reqRes.Errors = errors
+			requestResults <- reqRes
+			wg.Add(-1)
```

Comment:
use `wg.Done()` here

Reply:
I prefer it as-is because (1) it is more explicit, and (2) because we do some math around the number of these live goroutines (we do `wg.Add(1)` elsewhere).